src/net/java/sip/communicator/plugin/reconnectplugin/PPReconnectWrapper.java

```diff
@@ -358,4 +373,93 @@ public String toString()
 
         return builder.toString();
     }
+
+    /**
+     * Schedules a reconnect if needed (if there is timer and connected
+     * interfaces and user request is not null).
+     * @param delay The delay to use when creating the reconnect task.
+     * @param pp the protocol provider that will be reconnected.
+     */
+    private ReconnectTask scheduleReconnectIfNeeded(
+        long delay, ProtocolProviderService pp)
+    {
+        final ReconnectTask task = new ReconnectTask();
+        task.delay = delay;
+
+        if (timer == null)
+        {
+            return null;
+        }
+
+        if (!anyConnectedInterfaces())
+        {
+            // There is no network, nothing to do, when
+            // network is back it will be scheduled to reconnect.
+            // This means we started unregistering while
+            // network was going down and meanwhile there
+            // were no connected interface, this happens
+            // when we have more than one connected
+            // interface and we got 2 events for down iface
+
+            return null;
+        }
+
+        if(logger.isInfoEnabled())
+            logger.info("Reconnect " + pp + " after " + task.delay + " ms.");
+
+        timer.schedule(task, task.delay);
+
+        return task;
+    }
+
+    /**
+     * The task executed by the timer when time for reconnect comes.
+     */
+    private class ReconnectTask
+        extends TimerTask
+    {
+        /**
+         * The delay with which was this task scheduled.
+         */
+        long delay;
+
+        /**
+         * The thread to execute this task.
+         */
+        private Thread thread = null;
+
+        /**
+         * Reconnects the provider.
+         */
+        @Override
+        public void run()
+        {
+            if(thread == null || !Thread.currentThread().equals(thread))
+            {
+                thread = new Thread(this);
```

Comment:
I'm thinking that new thread is not needed for the reconnect task. This thread should be dead already by the time a new reconnect is needed and I see `Timer` instance serving the purpose of this thread. What do you think?

Reply:
O you are right since we moved to the per-connection timer. Updating it now.